### DISCUSSION TITLE
`World.FromPerceptionAnchor` is expensive?

### DISCUSSION THREAD

**Discussion by austinbhale** [StereoKit/StereoKit#615] [Q&A]
Hi Nick, I've been profiling lately and noticed something peculiar with `World.FromPerceptionAnchor`. I'm on the HoloLens 2 so using the unbounded reference space. I have a stepper that calls this method every frame which results in rather large spikes in CPU, about a 15-20% increase! Would you recommend only calling this method when [XR_TYPE_EVENT_DATA_REFERENCE_SPACE_CHANGE_PENDING](https://registry.khronos.org/OpenXR/specs/1.0/man/html/XrEventDataReferenceSpaceChangePending.html) is fired to indicate that a pending change in space is happening, so that we know when SK changes its origin?

**Comment by austinbhale** [comment]
I ended up polling [XR_TYPE_EVENT_DATA_REFERENCE_SPACE_CHANGE_PENDING](https://registry.khronos.org/OpenXR/specs/1.0/man/html/XrEventDataReferenceSpaceChangePending.html) like you do natively and now with https://github.com/StereoKit/StereoKit/pull/616, I can tell when the spatial anchor is found at the time of the event. Otherwise, if I tried to convert the spatial anchor to a pose, it outputted a pose with a zero position and ~almost identity-like quaternion except `w` was -1. When this happened, it was because of the failed call to `openxr_get_space`

Now that there's a way to specifically know when the space changes, the performance of this method is not a factor anymore :)

**Reply by maluoi** [reply]
I'm not entirely surprised that FromPerceptionAnchor is slow, but that seems like a lot! It is creating an OpenXR anchor and then dropping it right away iirc. Though I kinda always thought of it as stopgap until full anchor support is present.

I am currently working with origin related code, I'll make sure there's a nice easy event for when the reference space changes! I _may_ also fit in anchor support just yet too, but we'll see.

**Reply by austinbhale** [reply]
Looking forward to the origin and anchor support, high level access to that would feel so powerful!
